win32ss/gdi/eng/pdevobj.c

```diff
@@ -255,29 +334,62 @@ PDEVOBJ_pSurface(
     /* Increment reference count */
     GDIOBJ_vReferenceObjectByPointer(&ppdev->pSurface->BaseObject);
 
-    DPRINT("PDEVOBJ_pSurface() returning %p\n", ppdev->pSurface);
     return ppdev->pSurface;
 }
 
+VOID
+PDEVOBJ_vEnableDisplay(
+    _Inout_ PPDEVOBJ ppdev)
+{
+    BOOL assertVal;
+
+    if (!(ppdev->flFlags & PDEV_DISABLED))
+        return;
+
+    /* Try to enable display until success */
+    do
+    {
+        TRACE("DrvAssertMode(dhpdev %p, TRUE)\n", ppdev->dhpdev);
+        assertVal = ppdev->pfn.AssertMode(ppdev->dhpdev, TRUE);
+        TRACE("DrvAssertMode(dhpdev %p, TRUE) => %d\n", ppdev->dhpdev, assertVal);
+    } while (!assertVal);
```

Comment:
Looks like this could become an infinite loop if AssertMode always returns FALSE.

Reply:
According to MSDN : "A driver must return TRUE from a DrvAssertMode call with bEnable set to TRUE; that is, a driver cannot fail enabling a mode that was previously enabled."

Reply:
So, in which real-life cases an AssertMode call could "temporary" fail, and needs some retries until it can return TRUE ?